server/modules/protocol/MySQL/mysql_common.cc

```diff
@@ -1768,3 +1767,190 @@ void mxs_mysql_execute_kill_user(MXS_SESSION* issuer, const char* user, kill_typ
+                        (enum enum_session_state_type)mxs_leint_consume(&ptr);
+#if defined(SS_DEBUG)
+                    ss_dassert(type <= SESSION_TRACK_TRANSACTION_TYPE);
+#endif
+                    switch (type)
+                    {
+                        case SESSION_TRACK_STATE_CHANGE:
+                        case SESSION_TRACK_SCHEMA:
+                        case SESSION_TRACK_GTIDS:
+                            ptr += mxs_leint_consume(&ptr);
+                            break;
+                        case SESSION_TRACK_TRANSACTION_CHARACTERISTICS:
+                            mxs_leint_consume(&ptr); //length
+                            var_value = mxs_lestr_consume_dup(&ptr);
+                            gwbuf_add_property(buff, (char *)"trx_characteristics", var_value);
+                            MXS_FREE(var_value);
+                            break;
+                        case SESSION_TRACK_SYSTEM_VARIABLES:
+                            mxs_leint_consume(&ptr); //lenth 
+                            // system variables like autocommit, schema, charset ... 
+                            var_name = mxs_lestr_consume_dup(&ptr);
+                            var_value = mxs_lestr_consume_dup(&ptr);
+                            gwbuf_add_property(buff, var_name, var_value);
+                            MXS_DEBUG("SESSION_TRACK_SYSTEM_VARIABLES, name:%s, value:%s", var_name, var_value);
+                            MXS_FREE(var_name);
+                            MXS_FREE(var_value);
+                            break;
+                        case SESSION_TRACK_TRANSACTION_TYPE:
+                            mxs_leint_consume(&ptr); // length
+                            trx_info = mxs_lestr_consume_dup(&ptr);
+                            MXS_DEBUG("get trx_info:%s", trx_info);
+                            gwbuf_add_property(buff, (char *)"trx_state", trx_info);
+                            MXS_FREE(trx_info);
+                            break;
+                        default:
+                            ptr += mxs_leint_consume(&ptr);
+                            MXS_WARNING("recieved unexpecting session track type:%d", type);
+                            break;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/**
+ *  Check every packet type, if is ok packet then parse it
+ *  @param buff                 Buffer contain multi compelte packets
+ *  @param server_capabilities  Server capabilities
+ */
+void mxs_mysql_get_session_track_info(GWBUF *buff, uint32_t server_capabilities)
+{
+    size_t offset = 0;
+    uint8_t header_and_command[MYSQL_HEADER_LEN+1];
+
+    while(gwbuf_copy_data(buff, offset, MYSQL_HEADER_LEN+1, header_and_command) == (MYSQL_HEADER_LEN+1))
+    {
+        size_t packet_len = gw_mysql_get_byte3(header_and_command) + MYSQL_HEADER_LEN;
+        uint8_t cmd = header_and_command[MYSQL_COM_OFFSET];
+        if (packet_len > MYSQL_OK_PACKET_MIN_LEN && cmd == MYSQL_REPLY_OK)
```

Comment:
The following set of SQL statements should cause a false positive to be triggered.
```sql
CREATE TABLE t1(a VARCHAR(20), b INT, c INT, d INT);
INSERT INTO t1 VALUES ('', 100, 200, 300);
SELECT * FROM t1;
```
The result set row will be wrongly interpreted due to the fact that the MariaDB protocol is stateful. To work around this, the number of EOF packets could be tracked inside the protocol object stored in the DCB. To see whether a result set is in progress, we can just check if `num_eof_packets % 2 == 0`.